examples/lang/erc721/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_core::{
+    env::DefaultSrmlTypes,
+    memory::format,
+    storage,
+};
+use ink_lang::contract;
+use ink_model::EnvHandler;
+use ink_core::storage::alloc::Initialize as _;
+use ink_core::storage::alloc::AllocateUsing as _;
+use std::ops::Index;
+
+pub type Result<T, E> = core::result::Result<T, E>;
+
+contract! {
+    #![env = DefaultSrmlTypes]
+
+    // Event deposited when a token transfer occurs
+    event Transfer {
+        from: AccountId,
+        to: AccountId,
+        token_id: u32,
+    }
+
+    // Event deposited when an approval occurs
+    event Approval {
+        owner: AccountId,
+        to: AccountId,
+        token_id: u32,
+    }
+
+    /// The storage items for a typical ERC721 token implementation.
+    struct Erc721 {
+        /// The total supply.
+        token_owner: storage::HashMap<u32, AccountId>,
+        token_approvals: storage::HashMap<u32, AccountId>,
+        owned_tokens: storage::HashMap<AccountId, storage::Vec<u32>>,
+        all_tokens: storage::Vec<u32>,
+        name: storage::Vec<u8>,
+        symbol: storage::Vec<u8>,
+
+    }
+
+    impl Deploy for Erc721 {
+        fn deploy(&mut self, name: u8, symbol: u8) {
+            self.name.push(name);
+            self.symbol.push(symbol);
+        }
+    }
+
+    impl Erc721 {
+
+        pub(external) fn get_total_supply(&self) -> u32 {
+            let total_supply = self.all_tokens.len();
+            env.println(&format!("Erc20::total_supply = {:?}", total_supply));
+            total_supply
+        }
+        
+        pub(external) fn get_balance(&self, owner: AccountId) -> u32 {
+            let balance = self.balance_of(&owner);
+            env.println(&format!("Erc721::balance_of(owner = {:?}) = {:?}", owner, balance));
+            balance
+        }
+
+        pub(external) fn get_owner(&self, token_id: u32) -> AccountId {
+            let owner = self.owner_of(&token_id);
+            env.println(&format!("Erc721::owner_of(token = {:?}) = {:?}", token_id, owner));
+            owner
+        }
+
+    }
+
+    impl Erc721 {
+        
+        fn approve(&mut self, env: &mut EnvHandler<ink_core::env::ContractEnv<DefaultSrmlTypes>>, to: &AccountId, token_id: &u32) -> Result<(), &'static str> {
+            let caller = env.caller();
+
+            if caller == self.owner_of(token_id){
+
+                self.token_approvals.insert(*token_id, caller);
+
+                env.emit(Approval {
+                    owner: caller,
+                    to: *to,
+                    token_id: *token_id,
+                });
+
+                Ok(())
+
+            } else{
+                Err("not owner")
```

Comment:
I would prefer us sticking as closely as possible to the actual Erc721 standard. We can later always do it better since we are using Rust but for now we should try to stick to the given standard.